Formatters sometimes omit important info from parser error messages

<!--
    ESLint adheres to the [JS Foundation Code of Conduct](https://js.foundation/community/code-of-conduct).

    This template is for bug reports. If you are reporting a bug, please continue on. If you are here for another reason, please see below:

    1. To propose a new rule: http://eslint.org/docs/developer-guide/contributing/new-rules
    2. To request a change: http://eslint.org/docs/developer-guide/contributing/changes
    3. If you have any questions, please stop by our chatroom: https://gitter.im/eslint/eslint

    Note that leaving sections blank will make it difficult for us to troubleshoot and we may have to close the issue.
-->

**Tell us about your environment**

* **ESLint Version:** master
* **Node Version:** 8.4.0
* **npm Version:** 5.3.0

**What parser (default, Babel-ESLint, etc.) are you using?**

default

**Please show your full configuration:**

(no configuration)

**What did you do? Please include the actual source code causing the issue.**

I linted an invalid piece of JavaScript text which only consists of a single dot: `.`

```bash
$ echo '.' | eslint --stdin --no-eslintrc
```

**What did you expect to happen?**

I expected a parser error including the message `Unexpected token .`

**What actually happened? Please include the actual, raw output from ESLint.**

```
<text>
  1:1  error  Parsing error: Unexpected token
```

The `stylish` and `codeframe` formatters strip trailing `.` characters from report messages, presumably for stylistic reasons. However, this leads to confusing messages when the `.` is actually an important component of the message, rather than a period at the end of the sentence.

One way to address this would be to avoid removing the trailing `.` for `fatal` messages. Another solution would be to just stop removing trailing `.` characters entirely.

--- a/lib/formatters/codeframe.js
+++ b/lib/formatters/codeframe.js
@@ -47,7 +47,7 @@ function formatFilePath(filePath, line, column) {
  */
 function formatMessage(message, parentResult) {
     const type = (message.fatal || message.severity === 2) ? chalk.red("error") : chalk.yellow("warning");
-    const msg = `${chalk.bold(message.message.replace(/\.$/, ""))}`;
+    const msg = message.message.search(/ \.$/) >= 0 ? `${chalk.bold(message.message)}` : `${chalk.bold(message.message.replace(/\.$/, ""))}`;
     const ruleId = message.fatal ? "" : chalk.dim(`(${message.ruleId})`);
     const filePath = formatFilePath(parentResult.filePath, message.line, message.column);
     const sourceCode = parentResult.output ? parentResult.output : parentResult.source;

--- a/lib/formatters/codeframe.js
+++ b/lib/formatters/codeframe.js
@@ -47,7 +47,7 @@ function formatFilePath(filePath, line, column) {
  */
 function formatMessage(message, parentResult) {
     const type = (message.fatal || message.severity === 2) ? chalk.red("error") : chalk.yellow("warning");
-    const msg = message.message.search(/ \.$/) >= 0 ? `${chalk.bold(message.message)}` : `${chalk.bold(message.message.replace(/\.$/, ""))}`;
+    const msg = `${chalk.bold(message.message.replace(/([^ ])\.$/, "$1"))}`;
     const ruleId = message.fatal ? "" : chalk.dim(`(${message.ruleId})`);
     const filePath = formatFilePath(parentResult.filePath, message.line, message.column);
     const sourceCode = parentResult.output ? parentResult.output : parentResult.source;

--- a/lib/formatters/stylish.js
+++ b/lib/formatters/stylish.js
@@ -65,7 +65,7 @@ module.exports = function(results) {
                     message.line || 0,
                     message.column || 0,
                     messageType,
-                    message.message.replace(/\.$/, ""),
+                    message.message.search(/ \.$/) >= 0 ? message.message : message.message.replace(/\.$/, ""),
                     chalk.dim(message.ruleId || "")
                 ];
             }),

--- a/lib/formatters/stylish.js
+++ b/lib/formatters/stylish.js
@@ -65,7 +65,7 @@ module.exports = function(results) {
                     message.line || 0,
                     message.column || 0,
                     messageType,
-                    message.message.search(/ \.$/) >= 0 ? message.message : message.message.replace(/\.$/, ""),
+                    message.message.replace(/([^ ])\.$/, "$1"),
                     chalk.dim(message.ruleId || "")
                 ];
             }),

--- a/tests/lib/formatters/codeframe.js
+++ b/tests/lib/formatters/codeframe.js
@@ -170,6 +170,26 @@ describe("formatter:codeframe", () => {
         });
     });
 
+    describe("when passed a message that ends with ' .'", () => {
+        const code = [{
+            filePath: "foo.js",
+            messages: [{
+                ruleId: "foo",
+                message: "Unexpected .",
+                severity: 2,
+                source: "foo"
+            }],
+            errorCount: 1,
+            warningCount: 0
+        }];
+
+        it("should return a string in the correct format (retaining the ' .')", () => {
+            const result = formatter(code);
+
+            assert.equal(stripAnsi(result), "error: Unexpected . (foo) at foo.js\n\n\n1 error found.");
+        });
+    });
+
     describe("when passed multiple messages", () => {
         const code = [{
             filePath: "foo.js",

--- a/tests/lib/formatters/stylish.js
+++ b/tests/lib/formatters/stylish.js
@@ -152,6 +152,31 @@ describe("formatter:stylish", () => {
         });
     });
 
+    describe("when passed a message that ends with ' .'", () => {
+        const code = [{
+            filePath: "foo.js",
+            errorCount: 0,
+            warningCount: 1,
+            fixableErrorCount: 0,
+            fixableWarningCount: 0,
+            messages: [{
+                message: "Unexpected .",
+                severity: 1,
+                line: 5,
+                column: 10,
+                ruleId: "foo"
+            }]
+        }];
+
+        it("should return a string in the correct format (retaining the ' .')", () => {
+            const result = formatter(code);
+
+            assert.equal(result, "\nfoo.js\n  5:10  warning  Unexpected .  foo\n\n\u2716 1 problem (0 errors, 1 warning)\n");
+            assert.equal(chalkStub.yellow.bold.callCount, 1);
+            assert.equal(chalkStub.red.bold.callCount, 0);
+        });
+    });
+
     describe("when passed a fatal error message", () => {
         const code = [{
             filePath: "foo.js",

